### ISSUE TITLE
System.AggregateException: 'Construction of the PE image failed with one or more errors.

### ISSUE THREAD

**Issue by HPXCoder** [Washi1337/AsmResolver#245]
## Describe the bug
Wenn writing a dll  i got this error  

## Screenshots
![555555555555](https://user-images.githubusercontent.com/31288493/150430775-4a369f17-2f7d-457d-a567-81af8a4c4c51.PNG)

## Platform
 - OS: Windows 10
 - AsmResolver Version: [e.g. 4.8.0]

**Comment by dr4k0nia** [comment]
It seems like this is an error on your side. Youre using a member that is not present in the module youre trying to write. Please make sure that you import or create the type in the current module before using it.

You can read about member references here: https://asmresolver.readthedocs.io/en/latest/dotnet/importing.html

**Comment by HPXCoder** [comment]
thank you very much your solution solved the first erro of type (delegate2)
do you have any solution for the secend error 
![555555555555](https://user-images.githubusercontent.com/31288493/150436375-81022c9f-8153-42c6-b086-5a68f4970e72.PNG)

this one :D
@dr4k0nia and great thanks for you bro 👍

**Comment by dr4k0nia** [comment]
Have you made sure the Enum type is imported into the module?

**Comment by HPXCoder** [comment]
> Have you made sure the Enum type is imported into the module?

sorry but how can i import it

**Comment by Washi1337** [comment]
This second error occurs when AsmResolver is trying to read and parse custom attribute signatures that reference types that are not present in the assembly, or contains invalid metadata in general. This is often the case with obfuscated binaries like yours probably is (judging from the name it is trying to resolve). By default, AsmResolver does **not** ignore errors in the metadata. You can enable this by providing an `EmptyErrorListener` in the module reader parameters. Please refer to the following article to know how this can be done:

https://asmresolver.readthedocs.io/en/latest/dotnet/advanced-module-reading.html#pe-image-reading-parameters

It could also be a bug, but it's hard to know or fix anything without you providing a way to reproduce this issue.

**Comment by HPXCoder** [comment]
@Washi1337 
Thank you very much but it is not in the read
its  n the write 
some programs protected contain different trademark global module type names, how can i edit it 💯

**Comment by Washi1337** [comment]
It is a reader error.  See https://github.com/Washi1337/AsmResolver/blob/master/src/AsmResolver.DotNet/Signatures/CustomAttributeArgumentReader.cs#L139. 
See also #235 for why you might only notice this upon writing the module. 

I don't really understand your second question. Could you rephrase it perhaps?

Also I still cannot reproduce the error you're getting. I cannot fix anything if I don't provide me with details on how to get to the same result (as the issue template asked you).

**Comment by HPXCoder** [comment]
@Washi1337  ok
i use the Asmresolver to delete the delegate calls from obfuscated dll i decompiled it and the dll have a protection method that change some methods into classes use the MulticastingDelegate i use a tool to remove this delegates and decrypt some of encrypted strings but when i try to save the module this error come to me thats only
and i use this tool https://github.com/ribthegreat99OrN0P/Agile.NET-Deobfuscator-Latest

**Comment by HPXCoder** [comment]
@Washi1337  i want to edit it but this error come threw 
i will appreciate if you could tell me the reason and how to fix it 👍

**Comment by HPXCoder** [comment]
`
var ctx = new Context() { Assembly = ModuleDefinition.FromFile(read, parameters).Assembly };
var runtimeVersion = ctx.Assembly.ManifestModule.OriginalTargetRuntime.Version;
                Console.WriteLine($"resolving assemblies from framework version : {runtimeVersion}");
                var resolver = new DotNetCoreAssemblyResolver(runtimeVersion);
                foreach (var reference in ctx.Assembly.ManifestModule.AssemblyReferences)
                {
                    resolver.Resolve(reference);
                }
                var imageBuilder = new ManagedPEImageBuilder();
                var factory = new DotNetDirectoryFactory(MetadataBuilderFlags.PreserveAll);
                imageBuilder.DotNetDirectoryFactory = factory;
                
                ctx.Assembly.ManifestModule.Write(Path.GetFileNameWithoutExtension(ctx.Path) + "-deob" + (ctx.Assembly.ManifestModule.IsILLibrary ? ".dll" : ".exe"), imageBuilder);
               `


this give me no error but also no file is saved

**Comment by Washi1337** [comment]
> @Washi1337 i want to edit it but this error come threw
i will appreciate if you could tell me the reason and how to fix it +1

I am not sure what you want to edit, nor can I really help you if you don't provide the steps to reproducible your error. It can be a bug in AsmResolver, it can be a bug in the third party tool that you are using, it can be something else. Without a reproducible example I cannot do much.

> this give me no error but when i edit on it and remove delegates from the TopLevelTypes give me the second error

Removing types when the type is used somewhere can brick the application or cause import errors. Make sure that you're only removing types that are not referenced any more.

**Comment by Washi1337** [comment]
Any update on this? 

I still have no way to reproduce this. Without sample code and sample input binary, I can't really fix anything.

**Comment by Washi1337** [comment]
Closed for inactivity.

**Comment by BlauFx** [comment]
I'm getting the same error but from doing something different. Not sure if I should create a new issue for this. Basically this happens if I try to run these two lines of code:
`var mdlDef = ModuleDefinition.FromFile("/home/oxygen/Downloads/a.exe");`
`mdlDef.Write("/home/oxygen/Downloads/b.exe", new ManagedPEImageBuilder(MetadataBuilderFlags.PreserveAll));`
I'm experiencing this issue on Linux. On Windows this works just fine (using also a windows path).
Using version 4.10.0

Error message:
Unhandled exception. System.AggregateException: Construction of the PE image failed with one or more errors. (Underlying enum type System.AttributeTargets could not be resolved. Assuming System.Int32 for custom attribute argument.)

**Comment by Washi1337** [comment]
@BlauFx if it works on Windows but you get assembly resolution errors on Linux, it may be an indication that you did not install a runtime that is compatible with the original runtime that the executable was targeting. This may mean for example that you'll need to install the mono runtime on Linux if you're trying to process .NET Framework assemblies.

**Comment by BlauFx** [comment]
Installing the mono runtime solved my issue, thanks.
